File: tests/r_parity/04_csdid.R
# StatsPAI CS-DiD parity (R side) -- Module 04.
#
# Reads data/04_csdid.csv (the StatsPAI mpdta replica) and runs
# did::att_gt + did::aggte(type="simple"). Tolerance: rel < 1e-3.
#
# Note: did::att_gt's default est_method = "dr" is doubly-robust; we
# use est_method = "reg" to match sp.callaway_santanna's outcome-
# regression doubly-robust variant on this DGP.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(did)
})

MODULE <- "04_csdid"

df <- read_csv_strict(MODULE)
# did::att_gt internally re-codes never-treated rows to first_treat
# = Inf and silently truncates if the column is integer. Coerce to
# numeric so the never-treated cohort is preserved.
df$first_treat <- as.numeric(df$first_treat)
df$lemp        <- as.numeric(df$lemp)
df$year        <- as.numeric(df$year)

set.seed(PARITY_SEED)
att_gt_fit <- did::att_gt(
  yname    = "lemp",
  tname    = "year",
  idname   = "countyreal",
  gname    = "first_treat",
  data     = df,
  control_group = "nevertreated",
  est_method    = "reg",
  bstrap        = FALSE,
  cband         = FALSE
)

agg_simple <- did::aggte(att_gt_fit, type = "simple",
                          bstrap = FALSE, cband = FALSE)

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "simple_ATT",
    estimate  = agg_simple$overall.att,
    se        = agg_simple$overall.se,
    ci_lo     = agg_simple$overall.att - qnorm(0.975) * agg_simple$overall.se,
    ci_hi     = agg_simple$overall.att + qnorm(0.975) * agg_simple$overall.se,
    n         = nrow(df)
  )
)

write_results(MODULE, rows,
              extra = list(estimator = "reg",
                           control_group = "nevertreated",
                           bstrap = FALSE))

File: tests/r_parity/04_csdid.py
"""StatsPAI CS-DiD parity (Python side) -- Module 04.

Dumps sp.datasets.mpdta() and runs sp.callaway_santanna with the
'reg' (outcome-regression) doubly-robust variant. The companion
04_csdid.R reads the same CSV and runs did::att_gt + did::aggte.

Tolerance: rel < 1e-3 (iterative estimator). The replica is a
deterministic seed=42 simulation calibrated to the published
mpdta neighbourhood.
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "04_csdid"


def main() -> None:
    df = sp.datasets.mpdta()
    dump_csv(df, MODULE)

    fit = sp.callaway_santanna(
        df, y="lemp", g="first_treat", t="year", i="countyreal",
        estimator="reg", control_group="nevertreated",
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE, side="py", statistic="simple_ATT",
            estimate=float(fit.estimate),
            se=float(fit.se),
            ci_lo=float(fit.ci[0]) if fit.ci is not None else None,
            ci_hi=float(fit.ci[1]) if fit.ci is not None else None,
            n=int(len(df)),
        )
    ]

    write_results(
        MODULE, "py", rows,
        extra={
            "estimator": "reg",
            "control_group": "nevertreated",
            "method": fit.method,
            "se_note": (
                "The simple-ATT point estimate matches R did::aggte and "
                "Stata csdid at rel < 1e-15. The analytic SE also matches "
                "the R/Stata no-bootstrap reference within the registered "
                "1% tolerance after including the control-regression "
                "uncertainty in the outcome-regression IF."
            ),
        },
    )


if __name__ == "__main__":
    main()
